infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java

```diff
@@ -115,13 +132,37 @@ public void apply(
       componentObjects = SelectorFilter.filter(componentObjects, k8sComponent.getSelector());
     }
 
-    estimateCommandsMachineName(workspaceConfig, k8sComponent, componentObjects);
+    List<PodData> podsData = getPodDatas(componentObjects);
+
+    estimateCommandsMachineName(workspaceConfig, k8sComponent, podsData);
+
+    if (k8sComponent.getMountSources()) {
+      applyProjectsVolumes(podsData, componentObjects);
+    }
 
     applyEntrypoints(k8sComponent.getEntrypoints(), componentObjects);
 
     k8sEnvProvisioner.provision(workspaceConfig, environmentType, componentObjects, emptyMap());
   }
 
+  private void applyProjectsVolumes(List<PodData> podsData, List<HasMetadata> componentObjects) {
+    PersistentVolumeClaim volumeClaim =
+        newPVC(PROJECTS_VOLUME_NAME, "ReadWriteOnce", defaultProjectPVCSize);
+    componentObjects.add(volumeClaim);
+    for (PodData podData : podsData) {
+      // skip pods without containers
+      if (podData.getSpec() == null) {
```

Comment:
I wonder why do we need this check, do we really have pods without spec? Did you faced some issue because I expect that parser will throw an exception since such pod is not a valid.